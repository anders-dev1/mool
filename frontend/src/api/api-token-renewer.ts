import {AccessTokenClient, RenewAccessTokenCommand } from "./api-client";
import {store} from "../redux/store";
import {sessionActions} from "../redux/actions/session.action";

export async function RenewToken() {
    let refreshToken = localStorage.getItem("refresh-token");
    if (!refreshToken){
        return Promise.reject('There was no refresh token in storage.');
    }

    // use refresh token to get new access token
    let client = new AccessTokenClient();
    let request = new RenewAccessTokenCommand();
    request.refreshToken = refreshToken;

    await client.renew(request).then(
        result => {
            localStorage.setItem("access-token", result.accessToken);
        },
        error => {
            if (error.status === 404){
                store.dispatch(sessionActions.renewTokenExpired());
            }
        }
    );
}