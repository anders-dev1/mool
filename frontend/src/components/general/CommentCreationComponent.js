import * as Yup from "yup";
import InputError from "./InputErrorComponent";
import {Form, Formik} from "formik";
import ProfilePictureComponent from "./ProfilePictureComponent";
import TextArea from "./TextAreaComponent";
import Button from "./ButtonComponent";
import React from "react";
import {commentActions} from "../../redux/actions/comment.actions";
import {withTheme} from "styled-components";
import {connect} from "react-redux";

class CommentCreation extends React.Component {
    render() {
        return (
            <Formik
                initialValues={{comment: ''}}
                validationSchema={Yup.object().shape({
                    comment: Yup.string()
                        .min(10, <InputError>Must be 10 or more characters</InputError>)
                        .required(<InputError>Required</InputError>)
                })}
                onSubmit={(values, {resetForm}) => {
                    this.props.addCommentToThread(this.props.threadId, values.comment);

                    // Would be better for the form to somehow subscribe to the success event and then reset.
                    // You could pass the resetForm as a callback to the action, but that
                    // ruins SoC by having the action dealing directly with UI components instead
                    // of through state.
                    resetForm();
                }}>
                {({
                      values,
                      errors,
                      touched,
                      handleChange,
                      handleBlur,
                      handleSubmit,
                      isSubmitting
                  }) => (
                    <Form onSubmit={handleSubmit}>
                        <div className={"flex items-start"}>
                            <ProfilePictureComponent isComment/>
                            <div className={"flex-auto ml1"}>
                                <TextArea
                                    name={"comment"}
                                    placeholder={"Write your input here..."}
                                    onChange={handleChange}
                                    onBlur={handleBlur}
                                    value={values.comment}
                                    data-testid={"commentCreationTextArea"}/>
                                {touched.comment && errors.comment}
                            </div>
                        </div>

                        <div className={"flex justify-end mt1"}>
                            <Button primary type="submit"
                                    processing={this.props.postingComment}
                                    data-testid={"commentCreationPostBtn"}>Post</Button>
                        </div>
                    </Form>
                )}
            </Formik>
        )
    }
}

function mapState(state, props) {
    const postingComment = state.threadsReducer.threadsById[props.threadId].postingComment;
    return {postingComment}
}

const actionCreators = {
    addCommentToThread: commentActions.addCommentToThread
}

const connectedComponent = withTheme(connect(mapState, actionCreators)(CommentCreation));
export {connectedComponent as CommentCreation};
export * from './CommentCreationComponent'