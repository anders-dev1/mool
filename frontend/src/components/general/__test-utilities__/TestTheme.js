export const testTheme = {
    textColor: "black",
    primaryColor: "#007BFF",
    primaryHover: "rgba(0,123,255,0.33)",
    successColor: "#28A745",
    errorColor: "#F02849",
    backgroundColor: "#F0F2F5",
    buttonTextColor: "#FFFFFF",
    dividerColor: "#C7CBD1",
    inputColor: "#99A5B0",
    spinnerColor: "#FFFFFF transparent transparent transparent",
    font: "Acumin Pro, Arial, sans-serif !important;"
};