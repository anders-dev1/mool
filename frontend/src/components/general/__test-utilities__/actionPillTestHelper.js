import {screen} from "@testing-library/react";

export function findActionPillByIcon(icon) {
    return screen.getByText(icon).closest('div');
}

export function findActionPillNumberByIcon(icon) {
    const pill = findActionPillByIcon(icon);
    const numberP = pill.lastChild.lastChild;
    return numberP.textContent;
}