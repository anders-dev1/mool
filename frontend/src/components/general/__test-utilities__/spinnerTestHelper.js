export function findSpinner(container){
    return container.getElementsByClassName('lds-ring')[0];
}