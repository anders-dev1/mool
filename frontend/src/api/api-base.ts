import fetchIntercept, {FetchInterceptorResponse} from 'fetch-intercept';
import {RenewToken} from "./api-token-renewer";

class Request {
    url: string;
    config: Config;
    constructor(url: string, config: Config) {
        this.url = url;
        this.config = config;
    }
}

const ignoredPaths: string[] = [
    "/api/accesstoken/renew",
    "/api/accesstoken/authenticate"
]

export class ApiBase {
    constructor() {
        let originalRequest: Request;
        const interceptor = fetchIntercept.register({
            request(url: string, config: Config): Promise<any[]> | any[] {
                if (ignoredPaths.some((path) => url.endsWith(path))){
                    // Calling this function removes this interceptor from the list of interceptors.
                    // This stops our interceptor from processing any longer in the current request chain.
                    interceptor();
                    return [url, config];
                }

                originalRequest = new Request(url, config);

                const token = localStorage.getItem('access-token');
                if (!token){
                    return [url, config];
                }

                const myHeaders: HeadersInit = new Headers();
                myHeaders.set('Accept', 'application/json');
                myHeaders.set('Content-Type', 'application/json');
                myHeaders.set('Authorization', "Bearer " + token);

                config.headers = myHeaders;

                return [url, config];
            },
            response(response: FetchInterceptorResponse): Promise<FetchInterceptorResponse> | any {
                if (response.status != 401) {
                    return response;
                }

                if (originalRequest.url.includes('token')) {
                    return response;
                }

                let refreshToken = localStorage.getItem("refresh-token");
                if (!refreshToken) {
                    return response;
                }

                // use refresh token to get new access token
                return RenewToken().then(
                    result => {
                        const accessToken = localStorage.getItem('access-token');

                        const myHeaders: HeadersInit = new Headers();
                        myHeaders.set('Authorization', "Bearer " + accessToken);
                        myHeaders.set('Accept', 'application/json');
                        myHeaders.set('Content-Type', 'application/json');
                        originalRequest.config.headers = myHeaders;

                        return fetch(originalRequest.url, originalRequest.config).then(
                            result => {
                                return result;
                            }
                        );
                    }
                );
            }
        });
    }
}

export interface Config {
    method: string;
    headers: HeadersInit;
}