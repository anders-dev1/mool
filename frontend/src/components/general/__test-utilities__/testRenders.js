import React from 'react'
import {render} from '@testing-library/react'
import {Provider} from 'react-redux'
import {setupStore, store} from "../../../redux/store";
import {testTheme} from "./TestTheme";
import {ThemeProvider} from "styled-components";

// This is used to set up the store for tests.
export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        testStore = setupStore(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({children}) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return {testStore, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}

// This is used with the react-test-renderer to render dom in tests.
export function internalsForRenderer(
    ui,
    {
        preloadedState = {},
        testStore = setupStore(preloadedState)
    } = {}
){
    function Wrapper(children) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return Wrapper(ui);
}