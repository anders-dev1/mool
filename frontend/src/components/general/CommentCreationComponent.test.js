import {internalsForRenderer, renderWithProviders} from "./__test-utilities__/testRenders";
import {CommentCreation} from "./CommentCreationComponent";
import renderer from "react-test-renderer";
import {fireEvent, act, screen} from "@testing-library/react";
import {CommentOrder, CommentsClient} from "../../api";

const validCommentContent = "This is a test comment";
const mockedThreadId = 1;
let commentCreateMock;
beforeEach(() => {
    commentCreateMock = jest.spyOn(CommentsClient.prototype, 'create')
        .mockImplementation(() => {
            // We need this delay for the "submit button when submitting should show loading spinner" test.
            // Otherwise, the loading spinner will never show.
            return new Promise((resolve) => setTimeout(resolve, 100));
        });

    // Must mock this because the updateCommentsForThread action is called when a comment is posted.
    jest.spyOn(CommentsClient.prototype, 'comments')
        .mockImplementation(() => Promise.resolve({comments: []}));
});

afterEach(() => {
    commentCreateMock.mockRestore();
});

function createState(){
    return {
        threadsReducer: {
            threadsById: {
                1: {
                    commentOrder: CommentOrder.OldestFirst,
                    postingComment: false,
                    numberOfComments: 0
                }
            }
        }
    }
}

function renderForm() {
    renderWithProviders(
        <CommentCreation threadId={mockedThreadId}/>,
        {preloadedState: createState()});

    const textArea = screen.getByRole("textbox");
    const submitButton = screen.getByRole("button");

    return {textArea, submitButton}
}

// The nested TextArea component needs a div passed as a forwardRef to figure out how tall it should be.
// This is not necessary for this test which is why we just pass a simple div through the createNodeMock option in
// react-test-renderer.
function createNodeMock(element) {
    return document.createElement('div');
}
test('renders correctly', () => {
    const internalDom = internalsForRenderer(
        <CommentCreation threadId={1}/>,
        {preloadedState: createState()});

    const tree = renderer.create(internalDom, {createNodeMock}).toJSON();
    expect(tree).toMatchSnapshot();
});

describe('submit button clicked,', () => {
    test('when form valid, should post form to backend with provided information', async () => {
        // Arrange
        const {textArea, submitButton} = renderForm();

        // Act
        await act(async () => {
            fireEvent.change(textArea, {target: {value: validCommentContent}});
            await new Promise(process.nextTick);

            fireEvent.click(submitButton);
            await new Promise(process.nextTick);
        });

        // Assert
        expect(commentCreateMock).toBeCalledWith(mockedThreadId, validCommentContent);
    });

    test('when form valid, should show loading spinner', async () => {
        // Arrange
        const {textArea, submitButton} = renderForm();

        // Act
        await act(async () => {
            fireEvent.change(textArea, {target: {value: validCommentContent}});
            await new Promise(process.nextTick);

            fireEvent.click(submitButton);
            await new Promise(process.nextTick);
        });

        // Assert
        const spinner = screen.getByTestId('spinner');
        expect(spinner).toHaveStyle('display: flex');
    });

    test('should reset form', async () => {
        // Arrange
        const {textArea, submitButton} = renderForm();

        // Act
        await act(async () => {
            fireEvent.change(textArea, {target: {value: validCommentContent}});
            await new Promise(process.nextTick);

            fireEvent.click(submitButton);
            await new Promise(process.nextTick);
        });

        // Assert
        expect(textArea.value).toEqual("");
    });

    test('when no comment content specified, should show validation error', async () => {
        // Arrange
        const {textArea, submitButton} = renderForm();

        // Act
        await act(async () => {
            fireEvent.click(submitButton);
            await new Promise(process.nextTick);
        });

        // Assert
        const element = screen.getByText("Required");
        expect(element).toBeTruthy();
    });

    test('when comment content shorter than 10 characters, should show validation error', async () => {
        // Arrange
        const {textArea, submitButton} = renderForm();

        // Act
        await act(async () => {
            fireEvent.change(textArea, {target: {value: "9chars000"}});
            await new Promise(process.nextTick);

            fireEvent.click(submitButton);
            await new Promise(process.nextTick);
        });

        // Assert
        const element = screen.getByText("Must be 10 or more characters");
        expect(element).toBeTruthy();
    });
});