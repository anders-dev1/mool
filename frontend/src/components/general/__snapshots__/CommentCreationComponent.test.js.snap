// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="flex items-start"
  >
    <img
      className="sc-gsnTZi ikquYD"
      src="http://placekitten.com/256/256"
    />
    <div
      className="flex-auto ml1"
    >
      <div
        className="sc-dkzDqf kQYKHC"
      >
        <textarea
          data-testid="commentCreationTextArea"
          name="comment"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Write your input here..."
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="flex justify-end mt1"
  >
    <button
      className="sc-gKXOVf ea-dSCL"
      data-testid="commentCreationPostBtn"
      type="submit"
    >
      <div
        className="sc-iBkjds dfLXjt"
        data-testid="childcontainer"
      >
        Post
      </div>
      <div
        className="sc-ftvSup jRdbpG"
        data-testid="spinner"
      >
        <div
          className="sc-hKMtZM yROcv lds-ring"
          size="1.5"
        >
          <div
            className="sc-eCYdqJ hoxuOs"
            size="1.5"
          />
          <div
            className="sc-eCYdqJ hoxuOs"
            size="1.5"
          />
          <div
            className="sc-eCYdqJ hoxuOs"
            size="1.5"
          />
          <div
            className="sc-eCYdqJ hoxuOs"
            size="1.5"
          />
        </div>
      </div>
    </button>
  </div>
</form>
`;
